### PR TITLE
Safe mediacodec bindings

### PR DESCRIPTION
**Pull request by zarik5**
Add `MediaCodec` and `MediaFormat` bindings with most of their methods. Crypto and DRM support is not implemented.

**Testing**
Only the decoder code path is tested. [sample](https://github.com/alvr-org/ALVR/blob/master/alvr/experiments/client/src/video_decoder/mediacodec.rs)

**Note:** This PR needs to be rebased after #213 is merged

### CODE REVIEW

File: ndk/src/media/media_codec.rs [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+        let name = CString::new(key).unwrap();
+        let mut out: ffi::size_t = 0;
+        unsafe { ffi::AMediaFormat_getSize(self.as_ptr(), name.as_ptr(), &mut out as _) }
+            .then(|| out as _)
+    }
+
+    pub fn get_buffer(&self, key: &str) -> Option<&[u8]> {
+        let name = CString::new(key).unwrap();
+        let mut out_buffer: *mut c_void = ptr::null_mut();
+        let mut out_size: ffi::size_t = 0;
```
Comment (MarijnS95) [inline ndk/src/media/media_codec.rs:79]: No need for the explicit typeS here.

File: ndk/src/media/media_codec.rs [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+    pub fn get_f32(&self, key: &str) -> Option<f32> {
+        let name = CString::new(key).unwrap();
+        let mut out = 0.0;
+        unsafe { ffi::AMediaFormat_getFloat(self.as_ptr(), name.as_ptr(), &mut out as _) }
+            .then(|| out)
+    }
+
+    pub fn get_usize(&self, key: &str) -> Option<usize> {
+        let name = CString::new(key).unwrap();
+        let mut out: ffi::size_t = 0;
```
Comment (MarijnS95) [inline ndk/src/media/media_codec.rs:71]: No need for the explicit type here.

File: ndk/src/media/media_codec.rs [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+    pub fn get_buffer(&self, key: &str) -> Option<&[u8]> {
+        let name = CString::new(key).unwrap();
+        let mut out_buffer: *mut c_void = ptr::null_mut();
+        let mut out_size: ffi::size_t = 0;
+        unsafe {
+            ffi::AMediaFormat_getBuffer(
+                self.as_ptr(),
+                name.as_ptr(),
+                &mut out_buffer as _,
+                &mut out_size as _,
```
Comment (MarijnS95) [inline ndk/src/media/media_codec.rs:85]: These will coerce, no need for `as _`.

**Review by MarijnS95 — CHANGES_REQUESTED** [review]
For starters, there are a lot of unnecessary `as _` casts and typed initializations. Please either be explicit about a cast or remove it altogether.

Adding `#![warn(trivial_casts)]` at the top of this file (hopefully we can do that repo-wide at some point) will help you track these down.

**Comment by MarijnS95** [comment]
Fwiw feel free to leave the warning at the top of the file for the meantime. I'll work out to combine that with #219.

**Comment by MarijnS95** [comment]
> ```
> cargo clippy -p ndk --target armv7-linux-androideabi --features api-level-30 --features media
> cargo clippy -p ndk --target aarch64-linux-android --features api-level-30 --features media
> ```

You may also use `x86_64`/`i686` as those have a different pointer type for `c_char` (`u8`), afaik `armv7` also has `i8` like `aarch64`?

**Comment by zarik5** [comment]
> You may also use x86_64/i686 as those have a different pointer type for c_char (u8), afaik armv7 also has i8 like aarch64?

Ok, tried and no extra warnings or errors.

**Comment by zarik5** [comment]
There are still some warnings because of unused imports when using api-level-24, not only from my code

**Comment by MarijnS95** [comment]
> There are still some warnings because of unused imports when using api-level-24, not only from my code

That is something we should perhaps build-test in the CI and fix? Mind opening a separate PR for that (or issue if nontrivial)?

**Comment by zarik5** [comment]
Can this PR be merged? The warnings can be addressed in a second moment.

**Review by dvc94ch — APPROVED** [review]

**Comment by dvc94ch** [comment]
LGTM, thanks @zarik5

**Comment by MarijnS95** [comment]
I wasn't actually okay with this yet, but that's what I get for being on holiday. Doc comments with double instead of triple slashes, [`get_` prefixes](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter), changelog in the past tense and no reference to this PR, inconsistent `use` statements that are unused when certain features are not enabled, `enum` that doesn't [derive common traits like `Eq`](https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits) but instead has a strange `matches!` (why aren't the flags modeled directly with a `#[repr(u32)]` enum?), and that's just the stylistic nits.

**Comment by dvc94ch** [comment]
Sometimes things are good enough. But agree that it would be better with the changes you mentioned.

**Comment by MarijnS95** [comment]
If crates are going to be maintained (or at least PRs merged) with a "good is good enough" mentality I'll have to step down as a maintainer, as I can't be cleaning up behind peoples back to uphold an acceptable feeling of responsibility towards said project.

That is, unless there's clear agreement and documentation that a PR is dragging on too long (in terms of discussions, not silence), and is best merged in preparation for followup changes to land through separate PRs.

**Comment by zarik5** [comment]
I'm sorry @MarijnS95, I realize that things were rushed. I tried picking up the style around but I did a poor job. The wrong doc comments were the most glaring mistake for me. I think it would really help a PR checklist (not necessarily in this repo, you could link to a good one in another repo or website).

**Comment by MarijnS95** [comment]
I don't think a checklist for standard style errors really exists or make sense; it should all be caught by (locally runnable) CI tools like `clippy`. In the grand scheme of things I'd like to fail on many more default-allow (rustc or clippy) lint warnings, but it takes a ginormous effort to get this project to pass stricter lints initially :no_mouth: 

We could however use a checklist for obvious things like requiring a changelog entry if a PR modifies public-facing API/behaviour.

**Comment by MarijnS95** [comment]
In any case, my post above links at least twice to https://rust-lang.github.io/api-guidelines/ - that's a good starting place if you're looking for more guidance.

**Pull request merged**